x-pack/auditbeat/module/system/package/package.go

```diff
@@ -0,0 +1,537 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package pkg
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/gob"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/OneOfOne/xxhash"
+	"github.com/gofrs/uuid"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/auditbeat/datastore"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/auditbeat/cache"
+	"github.com/elastic/go-sysinfo"
+	"github.com/elastic/go-sysinfo/types"
+)
+
+const (
+	moduleName    = "system"
+	metricsetName = "package"
+	namespace     = "system.audit.package"
+
+	redhat = "redhat"
+	debian = "debian"
+	darwin = "darwin"
+
+	dpkgStatusFile     = "/var/lib/dpkg/status"
+	homebrewCellarPath = "/usr/local/Cellar"
+
+	bucketName              = "package.v1"
+	bucketKeyPackages       = "packages"
+	bucketKeyStateTimestamp = "state_timestamp"
+
+	eventTypeState = "state"
+	eventTypeEvent = "event"
+)
+
+type eventAction uint8
+
+const (
+	eventActionExistingPackage eventAction = iota
+	eventActionPackageInstalled
+	eventActionPackageRemoved
+)
+
+func (action eventAction) String() string {
+	switch action {
+	case eventActionExistingPackage:
+		return "existing_package"
+	case eventActionPackageInstalled:
+		return "package_installed"
+	case eventActionPackageRemoved:
+		return "package_removed"
+	default:
+		return ""
+	}
+}
+
+func init() {
+	mb.Registry.MustAddMetricSet(moduleName, metricsetName, New,
+		mb.DefaultMetricSet(),
+		mb.WithNamespace(namespace),
+	)
+}
+
+// MetricSet collects data about the system's packages.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config    config
+	log       *logp.Logger
+	cache     *cache.Cache
+	bucket    datastore.Bucket
+	lastState time.Time
+	osFamily  string
+}
+
+// Package represents information for a package.
+type Package struct {
+	Name        string
+	Version     string
+	Release     string
+	Arch        string
+	License     string
+	InstallTime time.Time
+	Size        uint64
+	Summary     string
+	URL         string
+}
+
+// Hash creates a hash for Package.
+func (pkg Package) Hash() uint64 {
+	h := xxhash.New64()
+	h.WriteString(pkg.Name)
+	h.WriteString(pkg.InstallTime.String())
+	return h.Sum64()
+}
+
+func (pkg Package) toMapStr() common.MapStr {
+	mapstr := common.MapStr{
+		"name":        pkg.Name,
+		"version":     pkg.Version,
+		"installtime": pkg.InstallTime,
+	}
+
+	if pkg.Arch != "" {
```

Comment:
Nit: Couldn't you use `putIfNotEmpty` here as well?

Reply:
I could have indeed, and I briefly thought about it but shied away from copying the function. Looking at more of this type of code throughout the Beats codebase I think `putIfNotEmpty` should be added to `common.MapStr` so we can use it everywhere. I'll make a note to propose adding it when I have a minute.